### PR TITLE
feat: improve kubernetes discovery

### LINKED ISSUE #6661: docs: there is no doc about k8s discovery
### Current State

No doc about k8s discovery.

### Desired State

We should add the doc under https://github.com/apache/apisix/tree/master/docs/en/latest/discovery.

@zhixiongdu027 
Would you handle this task? Thanks!

### PR DESCRIPTION
**Pull request by zhixiongdu027**
### Description

Fix #6661 
1. add  en discovery/kubernetes.md
2. improve zh discovery/kubernetes.md
3. use ngx.shared.kubernets instead ngx.shared.discovery

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/cli/ngx_tpl.lua [hunk 1]
```diff
@@ -190,6 +190,10 @@ http {
     lua_shared_dict balancer-ewma-last-touched-at {* http.lua_shared_dict["balancer-ewma-last-touched-at"] *};
     lua_shared_dict etcd-cluster-health-check {* http.lua_shared_dict["etcd-cluster-health-check"] *}; # etcd health check
 
+    {% if enabled_discoveries["kubernetes"] then %}
+    lua_shared_dict tars {* http.lua_shared_dict["kubernetes"] *};
```
Comment (spacewander) [inline apisix/cli/ngx_tpl.lua:194]: wrong shdict name?

File: docs/en/latest/discovery/kubernetes.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+## Summary
+
+The kubernetes service discovery plugin list&watch real-time changes of kubernetes cluster v1.endpoints, \
```
Comment (spacewander) [inline docs/en/latest/discovery/kubernetes.md:26]:
```suggestion
The kubernetes service discovery lists & watches real-time changes of kubernetes cluster v1.endpoints,
```

File: docs/en/latest/discovery/kubernetes.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+then store its value into ngx.shared.kubernetes.\
```
Comment (spacewander) [inline docs/en/latest/discovery/kubernetes.md:27]:
```suggestion
then stores theirs value into APISIX.
```

File: docs/en/latest/discovery/kubernetes.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+Discovery plugin also provides a query interface in accordance with the _APISIX Discovery specification_
+
+## Configuration
+
+A detailed configuration for the kubernetes service discovery plugin is as follows:
```
Comment (spacewander) [inline docs/en/latest/discovery/kubernetes.md:32]:
```suggestion
A detailed configuration for the kubernetes service discovery is as follows:
```

File: docs/en/latest/discovery/kubernetes.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+If the kubernetes service discovery plugin runs inside a pod, you can use minimal configuration:
+
+```yaml
+discovery:
+  kubernetes: { }
+```
+
+If the kubernetes service discovery plugin runs outside a pod, you need to create or select a specified _ServiceAccount_,
```
Comment (spacewander) [inline docs/en/latest/discovery/kubernetes.md:87]:
```suggestion
If the kubernetes service discovery runs outside a pod, you need to create or select a specified _ServiceAccount_,
```

File: docs/en/latest/discovery/kubernetes.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+      host: # enter apiserver host value here
+      port: # enter apiServer port value here
+    client:
+      token: # enter serviceaccount token value here
+      #token_file: # enter file path here
+```
+
+## Interface
+
+the kubernetes service discovery plugin provides a query interface in accordance with the _APISIX Discovery specification_
```
Comment (spacewander) [inline docs/en/latest/discovery/kubernetes.md:104]:
```suggestion
the kubernetes service discovery provides a query interface in accordance with the _APISIX Discovery specification_
```

File: apisix/discovery/kubernetes/init.lua [hunk 7]
```diff
@@ -30,7 +30,11 @@ local util = require("apisix.cli.util")
 local local_conf = require("apisix.core.config_local").local_conf()
 local informer_factory = require("apisix.discovery.kubernetes.informer_factory")
 
-local endpoint_dict
+local endpoint_dict = ngx.shared.kubernetes
+if not endpoint_dict then
+    error("failed to get nginx shared dict: kubernetes, please check your APISIX version")
```
Comment (tokers) [inline apisix/discovery/kubernetes/init.lua:35]:
```suggestion
    error("failed to get lua_shared_dict: kubernetes, please check your APISIX version")
```

File: docs/en/latest/discovery/kubernetes.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+       #- default
+       #- ^my-[a-z]+$
+
+    # kubernetes discovery plugin support use label_selector
+    # for the expression of label_selector, please refer to https://kubernetes.io/docs/concepts/overview/working-with-objects/labels
+    label_selector: |-
+      first="a",second="b"
+```
+
+If the kubernetes service discovery plugin runs inside a pod, you can use minimal configuration:
```
Comment (spacewander) [inline docs/en/latest/discovery/kubernetes.md:80]: Strictly speaking, The kubernetes service discovery is not a plugin.

File: docs/en/latest/discovery/kubernetes.md (hunk 2 above)
Comment (tokers) [inline docs/en/latest/discovery/kubernetes.md:26]:
```suggestion
The [Kubernetes](https://kubernetes.io/) service discovery plugin list & watch real-time changes of [Endpoints](https://kubernetes.io/docs/concepts/services-networking/service/) resources in Kubernetes cluster, \
```

File: docs/en/latest/discovery/kubernetes.md (hunk 5 above)
Comment (tokers) [inline docs/en/latest/discovery/kubernetes.md:87]: I suggest giving the way to fetch the token value. For instance, if the token in the SA was specified in a Secret `default-token-abc`, then we can:

```shell
kubectl get secrets default-token-abc  -o jsonpath={.data.token} | base64 -d
```

File: docs/en/latest/discovery/kubernetes.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+## Q&A
+
+> Q: Why only support configuration token to access _Kubernetes ApiServer_ \
+> A: Usually, we will use three ways to complete the authentication of _Kubernetes ApiServer_:
+>
+>+ mTLS
+>+ token
+>+ basic authentication
+>
+> Because lua-resty-http does not currently support mTLS, and basic authentication is not recommended,\
```
Comment (tokers) [inline docs/en/latest/discovery/kubernetes.md:162]: I think lua-resty-http has the capability to support mTLS, see https://github.com/apache/apisix/blob/master/conf/config-default.yaml#L279-L287, we already support mTLS in the communication with ETCD.

**Comment by zhixiongdu027** [comment]
@spacewander @tokers 
Hi, all comment issues have been taken care of. please help review.
Many TKS.

File: docs/en/latest/discovery/kubernetes.md (hunk 9 above)
Comment (zhixiongdu027) [reply docs/en/latest/discovery/kubernetes.md:162]: TKS you remained,.
Maybe we can pretend we don't know and add mTLS in the future @_@

**Comment by zhixiongdu027** [comment]
@spacewander @tokers 
TKS review, ALL comment issues have been taken care of. please help reviews

**Comment by tao12345666333** [comment]
let me re-run CI.

**Review by spacewander — DISMISSED** [review]

**Comment by spacewander** [comment]
Record the CI failure with this issue: https://github.com/openresty/openresty-packaging/issues/79

File: docs/zh/latest/discovery/kubernetes.md [hunk 10]
```diff
@@ -115,6 +119,6 @@ service_name 的 pattern 如下:
 
 -------
 
-> Q: APISIX 是多进程模型, 是否意味着每个 APISIX 工作进程都会监听 Kubernetes v1.endpoints \
+> Q: APISIX 继承了Nginx 的多进程模型, 是否意味着每个 APISIX 工作进程都会监听 Kubernetes v1.endpoints \
```
Comment (leslie-tsang) [inline docs/zh/latest/discovery/kubernetes.md:122]:
```suggestion
> Q: APISIX 继承了 Nginx 的多进程模型, 是否意味着每个 APISIX 工作进程都会监听 Kubernetes v1.endpoints \
```

File: docs/zh/latest/discovery/kubernetes.md [hunk 11]
```diff
@@ -115,6 +119,6 @@ service_name 的 pattern 如下:
...
 > A: Kubernetes 服务发现插件只使用特权进程监听 Kubernetes v1.endpoints, 然后将结果存储\
-> 在 ngx.shared.dict 中, 业务进程是通过查询 ngx.shared.dict 来获取结果的
+> 在 ngx.shared.kubernetes 中, 工作进程是通过查询 ngx.shared.kubernetes 来获取结果的
```
Comment (leslie-tsang) [inline docs/zh/latest/discovery/kubernetes.md:124]:
```suggestion
> 在 ngx.shared.kubernetes 中, 工作进程是通过查询 ngx.shared.kubernetes 来获取结果
```

**Review by tao12345666333 — APPROVED** [review]

**Review by leslie-tsang — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Comment by spacewander** [comment]
@zhixiongdu027 
Would you like to update https://github.com/apache/apisix/blob/master/docs/en/latest/discovery.md#supported-discovery-registries? Thanks!

**Comment by zhixiongdu027** [comment]
image

It looks like we just need to insert a line here?

@spacewander

**Pull request merged**